File: stripe_terminal/android/src/main/kotlin/mek/stripeterminal/plugin/ReaderDelegatePlugin.kt
package mek.stripeterminal.plugin

import com.stripe.stripeterminal.external.callable.Cancelable
import com.stripe.stripeterminal.external.callable.HandoffReaderListener
import com.stripe.stripeterminal.external.callable.InternetReaderListener
import com.stripe.stripeterminal.external.callable.MobileReaderListener
import com.stripe.stripeterminal.external.callable.TapToPayReaderListener
import com.stripe.stripeterminal.external.models.BatteryStatus
import com.stripe.stripeterminal.external.models.DisconnectReason
import com.stripe.stripeterminal.external.models.Reader
import com.stripe.stripeterminal.external.models.ReaderDisplayMessage
import com.stripe.stripeterminal.external.models.ReaderEvent
import com.stripe.stripeterminal.external.models.ReaderInputOptions
import com.stripe.stripeterminal.external.models.ReaderSoftwareUpdate
import com.stripe.stripeterminal.external.models.TerminalException
import mek.stripeterminal.api.TerminalHandlersApi
import mek.stripeterminal.mappings.toApi
import mek.stripeterminal.runOnMainThread
import com.stripe.stripeterminal.external.callable.Callback

class ReaderDelegatePlugin(private val _handlers: TerminalHandlersApi) :
    MobileReaderListener, HandoffReaderListener, InternetReaderListener, TapToPayReaderListener {
    private var cancelableReconnect: Cancelable? = null
    private var cancelableUpdate: Cancelable? = null

    fun cancelReconnect(callback: Callback) {
        cancelableReconnect?.cancel(callback)
        if (cancelableReconnect == null) callback.onSuccess();
        cancelableReconnect = null
    }

    fun cancelUpdate(callback: Callback) {
        cancelableUpdate?.cancel(callback)
        if (cancelableUpdate == null) callback.onSuccess();
        cancelableUpdate = null
    }

    // ReaderListenable

    override fun onReportReaderEvent(event: ReaderEvent) = runOnMainThread {
        _handlers.readerReportEvent(event.toApi())
    }

    // ReaderReconnectionListener

    override fun onReaderReconnectStarted(reader: Reader, cancelReconnect: Cancelable, reason: DisconnectReason) =
        runOnMainThread {
            this.cancelableReconnect = cancelReconnect
            _handlers.readerReconnectStarted(reader.toApi(), reason.toApi())
        }

    override fun onReaderReconnectFailed(reader: Reader) = runOnMainThread {
        cancelableReconnect = null
        _handlers.readerReconnectFailed(reader.toApi())
    }

    override fun onReaderReconnectSucceeded(reader: Reader) = runOnMainThread {
        cancelableReconnect = null
        _handlers.readerReconnectSucceeded(reader.toApi())
    }
    // ReaderDisconnectListener

    override fun onDisconnect(reason: DisconnectReason) = runOnMainThread {
        _handlers.disconnect(reason.toApi())
    }

    // MobileReaderListener

    override fun onRequestReaderDisplayMessage(message: ReaderDisplayMessage) = runOnMainThread {
        _handlers.readerRequestDisplayMessage(message.toApi())
    }

    override fun onRequestReaderInput(options: ReaderInputOptions) = runOnMainThread {
        _handlers.readerRequestInput(options.options.mapNotNull { it.toApi() })
    }

    override fun onBatteryLevelUpdate(
        batteryLevel: Float,
        batteryStatus: BatteryStatus,
        isCharging: Boolean
    ) = runOnMainThread {
        _handlers.readerBatteryLevelUpdate(
            batteryLevel = batteryLevel.toDouble(),
            batteryStatus = batteryStatus.toApi(),
            isCharging = isCharging
        )
    }

    override fun onReportLowBatteryWarning() = runOnMainThread {
        _handlers.readerReportLowBatteryWarning()
    }

    override fun onReportAvailableUpdate(update: ReaderSoftwareUpdate) = runOnMainThread {
        _handlers.readerReportAvailableUpdate(update.toApi())
    }

    override fun onStartInstallingUpdate(update: ReaderSoftwareUpdate, cancelable: Cancelable?) =
        runOnMainThread {
            this.cancelableUpdate = cancelable
            _handlers.readerStartInstallingUpdate(update.toApi())
        }

    override fun onReportReaderSoftwareUpdateProgress(progress: Float) = runOnMainThread {
        _handlers.readerReportSoftwareUpdateProgress(progress.toDouble())
    }

    override fun onFinishInstallingUpdate(update: ReaderSoftwareUpdate?, e: TerminalException?) =
        runOnMainThread {
            cancelableUpdate = null
            _handlers.readerFinishInstallingUpdate(update?.toApi(), e?.toApi())
        }
}

File: mek_stripe_terminal/ios/mek_stripe_terminal/Sources/mek_stripe_terminal/Plugin/ReaderDelegatePlugin.swift
import Foundation
import StripeTerminal

class ReaderDelegatePlugin: NSObject, ReaderDelegate, MobileReaderDelegate, InternetReaderDelegate, TapToPayReaderDelegate {
    
    private let _handlers: TerminalHandlersApi
    private var cancellableReconnection: Cancelable?
    private var cancellableUpdate: Cancelable?

    init(_ handlers: TerminalHandlersApi) {
        self._handlers = handlers
    }
    
    func cancelReconnection() async throws {
        try await cancellableReconnection?.cancel()
        cancellableReconnection = nil
    }
    
    func cancelUpdate() async throws {
        try await cancellableUpdate?.cancel()
        cancellableUpdate = nil
    }
    
    // ReaderDelegate methods
    
    func reader(_ reader: Reader, didDisconnect reason: DisconnectReason) {
        DispatchQueue.main.async {
            self._handlers.disconnect(reason: reason.toApi())
        }
    }
    
    func reader(_ reader: Reader, didStartReconnect cancelable: Cancelable, disconnectReason: DisconnectReason) {
        self.cancellableReconnection = cancelable
        DispatchQueue.main.async {
            self._handlers.readerReconnectStarted(reader: reader.toApi(), reason: disconnectReason.toApi())
        }
    }
    
    func readerDidSucceedReconnect(_ reader: Reader) {
        cancellableReconnection = nil
        DispatchQueue.main.async {
            self._handlers.readerReconnectSucceeded(reader: reader.toApi())
        }
    }
    
    // MobileReaderDelegate methods
    
    func reader(_ reader: Reader, didReportAvailableUpdate update: ReaderSoftwareUpdate) {
        DispatchQueue.main.async {
            self._handlers.readerReportAvailableUpdate(update: update.toApi())
        }
    }
    
    func reader(_ reader: Reader, didStartInstallingUpdate update: ReaderSoftwareUpdate, cancelable: Cancelable?) {
        cancellableUpdate = cancelable
        DispatchQueue.main.async {
            self._handlers.readerStartInstallingUpdate(update: update.toApi())
        }
    }
    
    func reader(_ reader: Reader, didReportReaderSoftwareUpdateProgress progress: Float) {
        DispatchQueue.main.async {
            self._handlers.readerReportSoftwareUpdateProgress(progress: Double(progress))
        }
    }
    
    func reader(_ reader: Reader, didFinishInstallingUpdate update: ReaderSoftwareUpdate?, error: (any Error)?) {
        cancellableUpdate = nil
        let exception = (error as? NSError)?.toApi()
        DispatchQueue.main.async {
            self._handlers.readerFinishInstallingUpdate(
                update: update?.toApi(),
                exception: exception
            )
        }
    }
    
    func reader(_ reader: Reader, didRequestReaderInput inputOptions: ReaderInputOptions = []) {
        DispatchQueue.main.async {
            self._handlers.readerRequestInput(options: inputOptions.toApi())
        }
    }
    
    func reader(_ reader: Reader, didRequestReaderDisplayMessage displayMessage: ReaderDisplayMessage) {
        DispatchQueue.main.async {
            self._handlers.readerRequestDisplayMessage(message: displayMessage.toApi())
        }
    }
        
    func reader(_ reader: Reader, didReportReaderEvent event: ReaderEvent, info: Dictionary<AnyHashable, Any>?) {
        DispatchQueue.main.async {
            self._handlers.readerReportEvent(event: event.toApi())
        }
    }
    
    func reader(_ reader: Reader, didReportBatteryLevel batteryLevel: Float, status: BatteryStatus, isCharging: Bool) {
        DispatchQueue.main.async {
            self._handlers.readerBatteryLevelUpdate(
                batteryLevel: Double(batteryLevel),
                batteryStatus: status.toApi(),
                isCharging: isCharging
            )
        }
    }
    
    func readerDidReportLowBatteryWarning(_ reader: Reader) {
        DispatchQueue.main.async {
            self._handlers.readerReportLowBatteryWarning()
        }
    }
    
    // TapToPayReaderDelegate methods
    
    func tapToPayReader(_ reader: Reader, didStartInstallingUpdate update: ReaderSoftwareUpdate, cancelable: Cancelable?) {
        self.reader(reader, didStartInstallingUpdate: update, cancelable: cancelable)
    }
    
    func tapToPayReader(_ reader: Reader, didReportReaderSoftwareUpdateProgress progress: Float) {
        self.reader(reader, didReportReaderSoftwareUpdateProgress: progress)
    }
    
    func tapToPayReader(_ reader: Reader, didFinishInstallingUpdate update: ReaderSoftwareUpdate?, error: (any Error)?) {
        self.reader(reader, didFinishInstallingUpdate: update, error: error)
    }
    
    func tapToPayReader(_ reader: Reader, didRequestReaderInput inputOptions: ReaderInputOptions = []) {
        self.reader(reader, didRequestReaderInput: inputOptions)
    }
    
    func tapToPayReader(_ reader: Reader, didRequestReaderDisplayMessage displayMessage: ReaderDisplayMessage) {
        self.reader(reader, didRequestReaderDisplayMessage: displayMessage)
    }
    
    func tapToPayReaderDidAcceptTermsOfService(_ reader: Reader) {
        DispatchQueue.main.async {
            self._handlers.readerAcceptTermsOfService()
        }
    }
}
